File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Soda/Joins/Typed/JointEqualsIdentityTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.Tests.Common.Soda.Joins.Typed;

namespace Db4objects.Db4o.Tests.Common.Soda.Joins.Typed
{
	public class JointEqualsIdentityTestCase : AbstractDb4oTestCase
	{
		public class TestSubject
		{
			public string _name;

			public JointEqualsIdentityTestCase.TestSubject _child;

			public TestSubject(string name, JointEqualsIdentityTestCase.TestSubject child)
			{
				_name = name;
				_child = child;
			}
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Store()
		{
			JointEqualsIdentityTestCase.TestSubject subjectA = new JointEqualsIdentityTestCase.TestSubject
				("A", null);
			JointEqualsIdentityTestCase.TestSubject subjectB = new JointEqualsIdentityTestCase.TestSubject
				("B", subjectA);
			JointEqualsIdentityTestCase.TestSubject subjectC = new JointEqualsIdentityTestCase.TestSubject
				("C", subjectA);
			Store(subjectA);
			Store(subjectB);
			Store(subjectC);
		}

		public virtual void TestJointEqualsIdentity()
		{
			JointEqualsIdentityTestCase.TestSubject child = RetrieveChild();
			IQuery query = NewQuery(typeof(JointEqualsIdentityTestCase.TestSubject));
			IConstraint constraint = query.Descend("_name").Constrain("B").Equal();
			constraint.And(query.Descend("_child").Constrain(child).Identity());
			Assert.AreEqual(1, query.Execute().Count);
		}

		private JointEqualsIdentityTestCase.TestSubject RetrieveChild()
		{
			IQuery query = NewQuery(typeof(JointEqualsIdentityTestCase.TestSubject));
			query.Descend("_child").Constrain(null);
			return (JointEqualsIdentityTestCase.TestSubject)query.Execute().Next();
		}
	}
}

File: db4oj.tests/src/com/db4o/db4ounit/common/soda/joins/typed/JointEqualsIdentityTestCase.java
package com.db4o.db4ounit.common.soda.joins.typed;

import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class JointEqualsIdentityTestCase extends AbstractDb4oTestCase {

	public static class TestSubject {
		public String _name;
		public TestSubject _child;

		public TestSubject(String name, TestSubject child) {
			_name = name;
			_child = child;
		}
	}

	protected void store() throws Exception {
		TestSubject subjectA = new TestSubject("A", null);
		TestSubject subjectB = new TestSubject("B", subjectA);
		TestSubject subjectC = new TestSubject("C", subjectA);
		store(subjectA);
		store(subjectB);
		store(subjectC);
	}
	
	public void testJointEqualsIdentity() {
		TestSubject child = retrieveChild();
		Query query = newQuery(TestSubject.class);
		Constraint constraint = query.descend("_name").constrain("B").equal();
		constraint.and(query.descend("_child").constrain(child).identity());
		Assert.areEqual(1, query.execute().size());
	}

	private TestSubject retrieveChild() {
		Query query = newQuery(TestSubject.class);
		query.descend("_child").constrain(null);
		return (TestSubject) query.execute().next();
	}
}
